services/blockchain-indexer/shared/dataService/pos/validators.js

```diff
@@ -276,63 +276,70 @@ const updateValidatorListEveryBlock = () => {
 	const updateValidatorCacheListener = async (eventType, data) => {
 		const updatedValidatorAddresses = [];
 		const [block] = data.data;
-		if (block && block.transactions && Array.isArray(block.transactions)) {
-			block.transactions.forEach(tx => {
-				if (tx.module === MODULE.POS) {
-					if (tx.command === COMMAND.REGISTER_VALIDATOR) {
-						updatedValidatorAddresses
-							.push(getLisk32AddressFromPublicKey(tx.senderPublicKey));
-					} else if (tx.command === COMMAND.STAKE) {
-						// TODO: Verify
-						tx.params.stakes
-							.forEach(stake => updatedValidatorAddresses.push(stake.validatorAddress));
-					}
-				}
-			});
-
-			// TODO: Validate the logic if there is need to update validator cache on (un-)stake tx
-			if (updatedValidatorAddresses.length) {
-				const updatedValidatorAccounts = await business
-					.getPosValidators({ addresses: updatedValidatorAddresses });
-
-				updatedValidatorAccounts.forEach(validator => {
-					const validatorIndex = validatorList.findIndex(acc => acc.address === validator.address);
-
-					if (eventType === EVENT_DELETE_BLOCK && validatorIndex !== -1) {
-						// Remove validator from list when deleteBlock event contains validator registration tx
-						validatorList.splice(validatorIndex, 1);
-					} else if (validatorIndex === -1) {
-						// Append to validator list on newBlock event, if missing
-						validatorList.push(validator);
-					} else {
-						// Re-assign the current validator status before updating the validatorList
-						// Validator status can change only at the beginning of a new round
-						const { status } = validatorList[validatorIndex];
-						validatorList[validatorIndex] = { ...validator, status };
+		try {
+			if (block && block.transactions && Array.isArray(block.transactions)) {
+				block.transactions.forEach(tx => {
+					if (tx.module === MODULE.POS) {
+						if (tx.command === COMMAND.REGISTER_VALIDATOR) {
+							updatedValidatorAddresses
+								.push(getLisk32AddressFromPublicKey(tx.senderPublicKey));
+						} else if (tx.command === COMMAND.STAKE) {
+							// TODO: Verify
+							tx.params.stakes
+								.forEach(stake => updatedValidatorAddresses.push(stake.validatorAddress));
+						}
 					}
 				});
 
-				// Rank is impacted only when a validator gets (un-)voted
-				await computeValidatorRank();
-			}
+				// TODO: Validate the logic if there is need to update validator cache on (un-)stake tx
+				if (updatedValidatorAddresses.length) {
+					const updatedValidatorAccounts = await business
+						.getPosValidators({ addresses: updatedValidatorAddresses });
+
+					updatedValidatorAccounts.forEach(validator => {
+						const validatorIndex = validatorList
+							.findIndex(acc => acc.address === validator.address);
+
+						if (eventType === EVENT_DELETE_BLOCK && validatorIndex !== -1) {
+							// Remove validator from list when
+							// deleteBlock event contains validator registration tx
+							validatorList.splice(validatorIndex, 1);
```

Comment:
Would it be possible to make cached `validatorList` an instance of `Map` or `Object` with `address` as the key? It would bring time complexity of these operations from linear to constant.

Reply:
Discussed offline.